### DISCUSSION TITLE
EBSD reconstruction from dream3D INL file

### DISCUSSION THREAD

**Discussion by fuyao0610** [idaholab/moose#25180] [Q&A Modules: Phase field]
Hello all,

I encountered some difficulty in using the EBSDReader. The EBSD file was prepared using dream3D. Before it is read by moose, I also generated a IPF color map in dream3D to check if the grains are correctly shown in paraview (figure named IPF color), to make sure the data processing is correct.

Using the input script given in the moose named EulerAngleVariables2RGBAux.i and changing the EBSD file name, I try to read it in MOOSE, but I got the following results (shown in figure 'rgb') after opening the output from paraview.  The grain structures are generally correct but there are a few streaks. 

I also attached the EBSD file generated from dream3D (316INL.txt).  Is there some setting in Moose input script that I didn't set up correctly?

<img width="492" alt="IPFcolor" src="https://github.com/idaholab/moose/assets/24552518/cc40b4d7-1d7e-4310-9314-3058085aaf7b">
<img width="661" alt="rgb" src="https://github.com/idaholab/moose/assets/24552518/3259fb82-9d14-4192-bbd8-a7def8baf4b6">

[316INL.txt](https://github.com/idaholab/moose/files/12317248/316INL.txt)

**Comment by lindsayad** [comment]
@dschwen I think maybe you have the most experience with this reader?

**Comment by fuyao0610** [comment]
I am wondering if if you have time to take a look at this and provide some guidance?


Best,
Yao

**Comment by dschwen** [comment]
<s>Could you please share the MOOSE input as well?</s>

**Reply by dschwen** [reply]
It looks like the grain structure itself i read correctly:
![image](https://github.com/idaholab/moose/assets/202302/3740fe0b-2957-455b-ac01-8d1a0ead9f79)

**Reply by dschwen** [reply]
The averaged Euler angles for each grain also look fine. (this is Phi)
![image](https://github.com/idaholab/moose/assets/202302/d003728a-f739-4588-aa61-058976ea5497)

**Reply by dschwen** [reply]
Getting the non averaged (point) data looks fine to me as well:

![image](https://github.com/idaholab/moose/assets/202302/7f17a10d-1d5c-4264-888e-755be56c21bc)

Here I'm showing phi1, which is returned in degrees. an angle of 360deg is equivalent to an angle of 0deg, and orientations close to the poles can be close together an still show large precessions,  hence the apparent noise.

**Reply by dschwen** [reply]
My suspicion here is that everything is fine. The problem here is that you are looking at RGB wrong. That AuxKernel will actually output a 24bit number that corresponds to a value from 0x000000 to 0xffffff whci encodes a full RGB color value with 8bits per component. 

![image](https://github.com/idaholab/moose/assets/202302/6585ca91-0b5c-431c-8813-8ac95c323675)

Ok, I think I know what's going on. It looks like tehre is an overflow in that calculation. 

By the way the above viz can be made with paraview using a Calculator filter:

```
(rgb%256)*kHat + ((rgb/256)%256)*jHat + ((rgb/256/256)%256)*iHat
```

and by unticking _Map scalars_

**Reply by dschwen** [reply]
Let me try and fix this...

**Reply by dschwen** [reply]
No that's not it. Then the error must be in `Euler2RGB.C`. That code is nasty. At least this is only a visualization issue, not a reader issue. All the data looks like it's read correctly.

**Reply by fuyao0610** [reply]
Thanks for taking the time to look at it. If I understand correctly, the euler angles are correct but not the rgb color?  Also can I ask for the input script template because the one I used didn't get the ebsd_grains, phi_prime and phi_average at the output

**Reply by dschwen** [reply]
Yeah, here you go
[yao.i.txt](https://github.com/idaholab/moose/files/12479967/yao.i.txt)

**Comment by fuyao0610** [comment]
Sure, here it is.

[Mesh]
  [ebsd_mesh]
    type = EBSDMeshGenerator
    filename = 316INL.txt
  []
[]

[UserObjects]
  [ebsd]
    type = EBSDReader
  []
[]

[AuxVariables]
  [phi1]
    family = MONOMIAL
    order = CONSTANT
  []
  [phi]
    family = MONOMIAL
    order = CONSTANT
  []
  [phi2]
    family = MONOMIAL
    order = CONSTANT
  []

  [phase]
    order = CONSTANT
    family = MONOMIAL
  []
  [symmetry]
    order = CONSTANT
    family = MONOMIAL
  []

  [rgb]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  [phi1_aux]
    type = EBSDReaderPointDataAux
    variable = phi1
    ebsd_reader = ebsd
    data_name = phi1
    execute_on = initial
  []
  [phi_aux]
    type = EBSDReaderPointDataAux
    variable = phi
    ebsd_reader = ebsd
    data_name = phi
    execute_on = initial
  []
  [phi2_aux]
    type = EBSDReaderPointDataAux
    variable = phi2
    ebsd_reader = ebsd
    data_name = phi2
    execute_on = initial
  []

  [phase_aux]
    type = EBSDReaderPointDataAux
    variable = phase
    ebsd_reader = ebsd
    data_name = phase
    execute_on = initial
  []

  [symmetry_aux]
    type = EBSDReaderPointDataAux
    variable = symmetry
    ebsd_reader = ebsd
    data_name = symmetry
    execute_on = initial
  []

  [rgb]
    type = EulerAngleVariables2RGBAux
    variable = rgb
    phi1 = phi1
    phi = phi
    phi2 = phi2
    phase = phase
    symmetry = symmetry
    execute_on = initial
  []
[]

[Executioner]
  type = Steady
[]

[Problem]
  solve = false
  kernel_coverage_check = false
[]

[Outputs]
  exodus = true
#  show = 'rgb'
[]

Best,
Yao

**Comment by fuyao0610** [comment]
I couldn’t get this from the moose output, are you suggesting some changes of the input script?


Best,
Yao

**Reply by dschwen** [reply]
It's easier if you reply in thread rather than creating a new answer each time...

**Comment by fuyao0610** [comment]
Sure, I will reply in the thread in future
